src/frontend/src/utils/userNumber.ts

```diff
@@ -1,18 +1,186 @@
+};
+
+/** Read saved anchors from local storage, returning at most
+ * MAX_SAVED_ANCHORS */
+const getAnchors = (): bigint[] => {
+  const anchors = [];
+  const data = pruneAnchors(readAnchors());
+  for (const ix in data) {
+    anchors.push(BigInt(ix));
   }
+
+  // NOTE: this sort has nothing to do with the sorting we use to prune anchors.
+  // This sort here is only used to ensure users see a stable ordering of anchors.
+  anchors.sort();
+
+  return anchors;
+};
+
+/** Set the most recently used anchor */
+const setAnchorUsed = (anchor: bigint) => {
+  const anchors = readAnchors();
+  for (const ix in anchors) {
+    anchors[ix].unusedCount++;
+  }
+
+  const ix = anchor.toString();
+  // Here we try to be as non-destructive as possible and we keep potentially unknown
+  // fields
+  anchors[ix] = { ...anchors[ix], unusedCount: 0 };
+  writeAnchors(pruneAnchors(anchors));
+};
+
+/** Remove most unused anchors until there are MAX_SAVED_ANCHORS left */
+const pruneAnchors = (anchors: Anchors): Anchors => {
+  // this is equivalent to while(anchors.length > MAX)
+  // but avoids potential infinite loops if for some reason anchors can't
+  // be deleted
+  const extras = Math.max(Object.keys(anchors).length - MAX_SAVED_ANCHORS, 0);
+  for (let i = 0; i < extras; i++) {
+    const unused = mostUnused(anchors);
+    if (unused !== undefined) {
+      delete anchors[unused];
+    }
+  }
+
+  return anchors;
+};
+
+/** Best effort to migrate potentially old localStorage data to
+ * the latest format */
+export const migrate = () => {
+  // Exit early if we've already migrated/stored new anchors
+  if (getAnchors().length !== 0) {
+    return;
+  }
+
+  // Nothing to do if no 'userNumber's are stored
+  const userNumberString = localStorage.getItem("userNumber");
+  if (userNumberString === null) {
+    return;
+  }
+
+  // Try to decode old data
+  const userNumber = parseUserNumber(userNumberString);
+  if (userNumber === null) {
+    // The item may have been written by someone else so we don't remove it
+    console.warn("Could not migrate unknown user number", userNumber);
+    return;
+  }
+
+  // Most likely the anchor was used before being written to local storage
+  setAnchorUsed(userNumber);
+};
+
+/** Figure out which anchor was used the least recently (if there are any) */
+const mostUnused = (anchors: Anchors): string | undefined => {
+  // First turn the { 123: { unusedCount: ... } } into [{ ix: 123, unusedCount: ... }],
+  // then (reverse) sort by unusedCount and return first element
+
+  const arr: { ix: string; unusedCount: number }[] = [];
+  for (const ix in anchors) {
+    arr.push({ ix, unusedCount: anchors[ix].unusedCount });
+  }
+
+  arr.sort((a, b) => b.unusedCount - a.unusedCount);
+
+  return arr[0]?.ix;
+};
+
+/* "Low-level" serialization functions to read and write anchors to local storage */
+
+const readAnchors = (): Anchors => {
+  const blob = localStorage.getItem("anchors");
```

Comment:
by reading `blob` I kind of except it to be a `Blob` while actually it's a `string`. Renaming the variable to e.g. `text` would be a bit more easy to read and intepret.